res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,739 @@
+// Pioneer-DDJ-400-script.js
+// ****************************************************************************
+// * Mixxx mapping script file for the Pioneer DDJ-400.
+// * Author: Warker, nschloe
+// * Forum: https://mixxx.org/forums/viewtopic.php?f=7&t=12113
+// * Wiki: https://www.mixxx.org/wiki/doku.php/pioneer_ddj-400
+//
+// Upstream MIDI spec:
+// * https://www.pioneerdj.com/-/media/pioneerdj/software-info/controller/ddj-400/ddj-400_midi_message_list_e1.pdf
+//
+//             Working:
+//                 * Mixer Section (Faders, EQ, Filter, Gain, Cue)
+//                 * Browsing and loading + Waveform zoom (shift)
+//                 * Jogwheels, Scratching, Bending
+//                 * cycle Temporange
+//                 * Beat Sync
+//                 * Beat Loop Mode
+//                 * Sampler Mode
+//
+//             Partially:
+//                 * Beatjump mode (no lighting, shift mode)
+//                 * PAD FX (only slots A-H, Q-P)
+//                 * Effect Section (without Beat FX left + Right - no equivalent function found)
+//                 * Output (lights)
+//
+//             Testing:
+//                 * Keyboard Mode (check pitch value)
+//                 * Keyshift Mode (check pitch value)
+//                 * Hot Cue Mode (including loops)
+//                 * Loop Section: Loop in / Out + Adjust, Call, Double, Half
+//                 * Effect Section (Beat FX left + Right - select the Effect Slot (not Effect BPM))
+//
+//             Not working/implemented:
+//                 * Channel & Crossfader Start
+//
+var PioneerDDJ400 = {};
+
+var LightsPioneerDDJ400 = {
+    beatFx: {
+        status: 0x94,
+        data1: 0x47,
+    },
+    deck1: {
+        vuMeter: {
+            status: 0xB0,
+            data1: 0x02,
+        },
+        playPause: {
+            status: 0x90,
+            data1: 0x0B,
+        },
+        shiftPlayPause: {
+            status: 0x90,
+            data1: 0x47,
+        },
+        cue: {
+            status: 0x90,
+            data1: 0x0C,
+        },
+        shiftCue: {
+            status: 0x90,
+            data1: 0x48,
+        },
+    },
+    deck2: {
+        vuMeter: {
+            status: 0xB0,
+            data1: 0x02,
+        },
+        playPause: {
+            status: 0x91,
+            data1: 0x0B,
+        },
+        shiftPlayPause: {
+            status: 0x91,
+            data1: 0x47,
+        },
+        cue: {
+            status: 0x91,
+            data1: 0x0C,
+        },
+        shiftCue: {
+            status: 0x91,
+            data1: 0x48,
+        },
+    },
+};
+
+// Save the Shift State
+PioneerDDJ400.shiftState = [0, 0];
+
+// JogWheel
+PioneerDDJ400.vinylMode = true;
+PioneerDDJ400.alpha = 1.0/8;
+PioneerDDJ400.beta = PioneerDDJ400.alpha/32;
+PioneerDDJ400.highspeedScale = 2;
```

Comment:
So I was trying to work out why high-speed search via SHIFT+JOGWHEEL wasn't working, and it seems to trace back to this value being too low.  At a highspeedScale of 2, using SHIFT+JOGWHEEL moves through the track _very slowly_, i.e. at way less than normal playback speed or even scratching speed.

Toying about with this, a value of 75 to 100 seems to produce much better behavior in terms of being able to move through the track quickly, however i've noticed this can be very touchy.  If you stop touching the top of the jogwheel before releasing SHIFT, the track can sometimes "race ahead" even faster than expected for a second or two before coming to a stop.  If I'm careful to release the SHIFT button before taking my finger off of the jogwheel, things behave as expected.

Anyone played with this at all yet, and should we consider increasing this value?

Reply:
After further experimentation I settled on a value of 20 for highspeedScale.  This seems to work fairly well.